lib/jxl/jpeg/dec_jpeg_data_writer.cc

```diff
@@ -490,42 +506,115 @@ bool EncodeDCTBlockSequential(const coeff_t* coeffs, HuffmanCodeTable* dc_huff,
   temp2 = coeffs[0];
   temp = temp2 - *last_dc_coeff;
   *last_dc_coeff = temp2;
-  temp2 = temp;
-  if (temp < 0) {
-    temp = -temp;
-    if (temp < 0) return false;
-    temp2--;
-  }
-  int dc_nbits = (temp == 0) ? 0 : (FloorLog2Nonzero<uint32_t>(temp) + 1);
+  temp2 = temp >> (8 * sizeof(coeff_t) - 1);
+  temp += temp2;
+  temp2 ^= temp;
+
+  int dc_nbits = (temp2 == 0) ? 0 : (FloorLog2Nonzero<uint32_t>(temp2) + 1);
   WriteSymbol<kOutputMode>(dc_nbits, dc_huff, bw);
+#if false
+  // If the input is corrupt, this could be triggered. Checking is
+  // costly though, so it makes more sense to avoid this branch.
+  // (producing a corrupt JPEG when the input is corrupt, instead
+  // of catching it and returning error)
   if (dc_nbits >= 12) return false;
-  if (dc_nbits > 0) {
-    WriteBits(bw, dc_nbits, temp2 & ((1u << dc_nbits) - 1));
-  }
-  int r = 0;
-  for (int k = 1; k < 64; ++k) {
-    if ((temp = coeffs[kJPEGNaturalOrder[k]]) == 0) {
-      r++;
-      continue;
-    }
-    if (temp < 0) {
-      temp = -temp;
-      if (temp < 0) return false;
-      temp2 = ~temp;
-    } else {
-      temp2 = temp;
-    }
-    while (r > 15) {
-      WriteSymbol<kOutputMode>(0xf0, ac_huff, bw);
-      r -= 16;
-    }
-    int ac_nbits = FloorLog2Nonzero<uint32_t>(temp) + 1;
-    if (ac_nbits >= 16) return false;
-    int symbol = (r << 4u) + ac_nbits;
-    WriteSymbol<kOutputMode>(symbol, ac_huff, bw);
-    WriteBits(bw, ac_nbits, temp2 & ((1 << ac_nbits) - 1));
-    r = 0;
+#endif
+  WriteBits(bw, dc_nbits, temp & ((1u << dc_nbits) - 1));
+  int16_t r = 0;
+
+  // Manually unfolded loop over kJPEGNaturalOrder[k] for k=1..63
+#define kloop(jpeg_natural_order_of_k)                            \
```

Comment:
how much speedup is the loop unrolling in itself?

Reply:
Without that unrolling (keeping everything else the same) it's about 3% slower, e.g. `21.74 MP/s [21.29, 21.80]` without unrolling vs `22.39 MP/s [21.96, 22.49]` with unrolling.